### ISSUE TITLE
How to import additional languages with webpack?

### ISSUE THREAD

**Issue by mtness** [mymth/vanillajs-datepicker#29]
Hi there,
Could anybody please give an example how to import the language dependencies via webpack, too? Am I missing something?

```
@import Datepicker from 'path/to/node_modules/vanillajs-datepicker/js/Datepicker.js';
==
import { Datepicker } from 'vanillajs-datepicker';
```

```
@import de from 'path/to/node_modules/vanillajs-datepicker/js/i18n/locales/de.js';
== 
???
```

Thanks in advance and best regards!

**Comment by ocarreterom** [comment]
https://mymth.github.io/vanillajs-datepicker/#/i18n?id=adding-langauages
```js
@import Datepicker from 'path/to/node_modules/vanillajs-datepicker/js/Datepicker.js';
@import es from 'path/to/node_modules/vanillajs-datepicker/js/i18n/locales/es.js';

Object.assign(Datepicker.locales, es);
```

**Comment by MathiasZaja** [comment]
> https://mymth.github.io/vanillajs-datepicker/#/i18n?id=adding-langauages
> 
> ```js
> @import Datepicker from 'path/to/node_modules/vanillajs-datepicker/js/Datepicker.js';
> @import es from 'path/to/node_modules/vanillajs-datepicker/js/i18n/locales/es.js';
> 
> Object.assign(Datepicker.locales, es);
> ```

Where do you put @import Datenpicker from ... and @import es from ...
In the <script> somwhere on html page, in some other external js ?
By the way the doc says import (without @)
I'm desperately confused.

**Comment by ocarreterom** [comment]
I remove `@` from my code.
This is an example project https://codesandbox.io/s/loving-varahamihira-b1g22?file=/src/index.js

**Comment by MathiasZaja** [comment]
Hi Oscar,

 

Thanks a lot for having provided an example of adding an additional language to vanillas-datenpicker.

It works like a charm in codesandbox; I can add any other language of my choice.

However I’m not able to reproduce it locally.

I’ve downloaded the zip file and reproduced the tree structure.

I added a folder named vanillajs-datepicker and Datenpicker.js and i18n/locales folder (and all languages .js in it) All that stuff extracted from vanillajs-datepicker-master.zip

When I open index.html just the input field appears but no events attached to it.

 

I must be missing something obvious.

Do I have to run a script (npm ?) or do I have to do something else to make it work.

 

I do not use any kind of framework (bootstrap etc).

I just want to use the calendar in a standard Web application.

 

Sorry for my luck of competence and kind regards.

**Comment by MathiasZaja** [comment]
Hi Oscar,

 

I finally understood where I was wrong.

Vanilla Datepicker uses JavaScript modules which work only via HTTP(s) so it cannot be run locally (via file:// protocol)

 

So my question would be:

Do you have any version of Vanilla Datepicker which runs with “standard” JavaScript (i.e. without modules) ?

 

Kind regards

**Comment by ocarreterom** [comment]
You need a development server. 
For example https://github.com/http-party/http-server

**Comment by MathiasZaja** [comment]
Hi Oscar,

 

Thanks for your message.

 

I can execute Vanilla Datepicker in Eclipse IDE.

However my problem is that my Web app has the Java / servlet architecture (with Apache & Tomcat) with all JavaScript executing on client side.

To integrate your Datepicker I would have to install a static server with nodejs, or similar, in parallel. That is risky and generally not recommended.

If I am wrong, please correct me.

 

So for now I’m rewriting your scripts to be able to run them on client.

Tough job.

What would be fantastic for me is to have the chance to install your Datepicker like I TinyMCE (https://www.tiny.cloud/)

It is Christmas :)

 

Kind regards

**Comment by mymth** [comment]
@MathiasZaja,

The package includes pre-built js and css files in the `dist` directory. You can refer to the "Using from Browser" section and the "for browser" part of the i18n - Adding languages section in the docs to lean how to use them.

The docs say to use CDN, but of course it's not necessary because the pre-built files are standalone scripts/stylesheets; they can be loaded from anywhere.
Since you already downloaded the package, you can pick the files that meet your need from the `dist` directory and put them into your servlet's assets directory.

And FYI, they work via file:// protocol too. You may want to try double-clicking the `demo/index.html` file in your downloaded package directory.

**Comment by MathiasZaja** [comment]
Thanks a lot for your help,
I followed your instructions i.e. included datepicker.js, datepicker.css and fr.js from the dist directory.

### Extracts of my html page:

<.head.>
  <.link media="screen" href="/path-to-css/datepicker.css" rel="stylesheet".>
  <.script src="/path-to-js/datepicker.js".><./script.>
<./head.>

<.head.>
  <.link media="screen" href="/path-to-css/datepicker.css" rel="stylesheet".>
  <.script src="/path-to-js/datepicker.js".><./script.>
<./head.>
<.div style="mystyle" class="myclass".>
  <.input name="foo" type="text".>
<./div.>
<.script.>
  const elem = document.querySelector('input[name="foo"]');
  const datepicker = new Datepicker(elem, {
    'todayBtn':true,
    'title' : 'Choose date',
    'language' : 'fr',
  });          
<./script.>
<.script src="/path-to-locales/fr.js".><./script.>

### Result
I can set almost any option I need except to change languge.
In DevTools I get the error: "Uncaught SyntaxError: Unexpected token 'export'" which points to line 5 of fr.js:
_1 /**
2  * French translation for bootstrap-datepicker
3  * Nico Mollet <nico.mollet@gmail.com>
4 */
**5 export default {**
6   fr: {
7    days: ["dimanche", "lundi", "mardi", "mercredi", "jeudi", "vendredi", "samedi"],_

What am I doing wrong ?
Thanks in advance and regards.
PS:
I added "." in the tags to make my code readable (I was not able to use "insert code")

**Comment by MathiasZaja** [comment]
Concerning the file:// protocol.
You are perfectly right, thanks for the correction.
I was introduced in error reading the introduction to javascript modules [https://javascript.info/modules-intro] who's author claimed the opposite. (I might haven't found the best article on this topic.) I reported your correction to the author.
Kind regards.

**Comment by mymth** [comment]
It looks like you copied the locale file from the `js/i18n/locales` directory. The pre-built locales are in `dist/js/locales`. If you replace the current `fr.js` with `dist/js/locales/fr.js`, the error should be gone.

You also need to move `<.script src="/path-to-locales/fr.js".><./script.>` to somewhere above your script code. (Right after `<.script src="/path-to-js/datepicker.js".><./script.>` would be the best place.) With your code, the Datepicker constructor is called before the the fr locale is loaded. So the constructor ignores the `language: 'fr'` option because the fr locale is not available at that point.

As for file://, they are right and you misunderstand. On browser, loading ECMAScript modules (ES6 modules) using the `import` statement doesn't work with file://, but works with http://. I guess it's because browser's JS runtime doesn't have local file access for security reason.

By building process, pre-built files are converted not to use export/import so that they can be loaded with `<script>` tag. This is why they work with file://.
I'd like to recommend comparing `js/i18n/locales/fr.js` and `dist/js/locales/fr.js` and see the difference.

**Comment by MathiasZaja** [comment]
Hi @mymth ,
Great thank you for your help and your patience, and sorry for my confiusion(s).
Your diagnostic is perfectly right: I copied the wrong language files.
I also put them in wrong place in my HTML: I misundertood the instruction in the documentation

> for browser
Load the locale files you need after Datepicker

Is building process done with ?

> npm install --save-dev vanillajs-datepicker 

That is something I tried at the very beginning but did not succeed. 
I got the following message from npm:

> 21 error 404 Not Found - GET https://registry.npmjs.org/vanillas-datepicker - Not found
22 error 404
23 error 404 'vanillas-datepicker@latest' is not in the npm registry.
24 error 404 You should bug the author to publish it (or use the name yourself!)

Not being familiar with npm I gave up.
However, thanks to you, I now do have all I need and I am not, and hopefully never will be, concerned with that aspect.
Now I can integrate your datepicker in my project (booking system which permanently deals with dates and times) and will deploy it in production.
I hope I will not have to disturb you anymore.
Once again thanks for wonderfull calendar tool and for your help.

PS:
I found in your demo (*.html & live-demo.js)

> el.addEventListener('change', onChnageDirection);

I imagine this mistyping (onChnage instead of onChange) has no impact since it seems to be systematic (in html & js).

**Comment by ocarreterom** [comment]
```diff
- vanillas-datepicker@latest
+ vanillajs-datepicker@latest
```

**Comment by MathiasZaja** [comment]
Hi @ocarreterom,
Sorry for mistyping. Originally I typed it correctly and got:

> 19 verbose stack Error: Refusing to install package with name "vanillajs-datepicker" under a package
19 verbose stack also called "vanillajs-datepicker". Did you name your project the same
19 verbose stack as the dependency you're installing?

However I do not have to deal with npm, right ?
I'd like to as a question about datepicker.css
Some items are set with color/background-color explicitely. Example:

> .datepicker-cell.range.focused {
    background-color: #cfcfcf;
}

Some others inherit from the html container (surrounding div).
In my case, in some user done interface choices, it renders date picker hardly readable. Here an example:
https://ubik.ch/DatepickerColorWhite.png
What would you advice ? Shall I try to fix it manually, editing the css ?
Do you have by chance some css (themes) which are "container independant" ?
Thanks for your help and kind regards.

**Comment by mymth** [comment]
For the info about the  error you got:
https://thisdavej.com/node-newbie-error-npm-refusing-to-install-package-as-a-dependency-of-itself/

npm is not necessary but makes updating the package easier.

The container options is meant for input field inside a scrollable div.

I think editing css to add something like below is the best way,
```css
.datepicker-picker {
  color: #4a4a4a;
}
```
but if you really have to avoid it, you should be able to do something like this:
```javascript
document.querySelectorAll('.datepicker-picker').forEach((el) => {
  el.style.color = '#4a4a4a';
});
```

If you have farther questions, please create a new issue.
I'd like to close this issue because the docs now include the answer to the OP's question.
(and your questions are totally off-topic now.)

**Comment by MathiasZaja** [comment]
Hi @mymth 
Thanks for your answer and advices.
You can close this issue.
Kind regards.
